src/rust_option.rs

```diff
@@ -30,517 +27,240 @@ impl<T: Sized> OptionTarget for &mut T {}
+            .as_mut_option()
         }
     }
 }
 
 #[cfg(feature = "alloc")]
 impl<'a> RustOption<&'a RustString> {
     pub fn from_option_string_ref(other: Option<&'a String>) -> Self {
-        let _: () = assert_option_safe::<&String>();
-        match other {
-            None => Self::new(),
-            Some(r) => Self {
-                inner: OptionInner::new(RustString::from_ref(r)),
-            },
+        unsafe {
+            core::mem::transmute::<RustOption<&String>, RustOption<&RustString>>(RustOption::from(
+                other,
+            ))
         }
     }
 
     pub fn into_option_string_ref(self) -> Option<&'a String> {
-        let _: () = assert_option_safe::<&String>();
-        match self.into_value() {
-            None => None,
-            Some(r) => Some(r.as_string()),
-        }
+        unsafe { core::mem::transmute::<RustOption<&RustString>, RustOption<&String>>(self) }
+            .into_option()
     }
 
-    pub fn as_option_string_ref(&mut self) -> &mut Option<&'a String> {
-        let _: () = assert_option_safe::<&String>();
-        unsafe { &mut *(self as *mut RustOption<&'a RustString> as *mut Option<&'a String>) }
+    pub fn as_option_string_ref_mut(&mut self) -> &mut Option<&'a String> {
+        unsafe { &mut *(self as *mut RustOption<&RustString> as *mut RustOption<&String>) }
+            .as_mut_option()
     }
 }
 
 #[cfg(feature = "alloc")]
 impl<'a> RustOption<&'a mut RustString> {
     pub fn from_option_string_mut(other: Option<&'a mut String>) -> Self {
-        let _: () = assert_option_safe::<&mut String>();
-        match other {
-            None => Self::new(),
-            Some(r) => Self {
-                inner: OptionInner::new(RustString::from_mut(r)),
-            },
+        unsafe {
+            core::mem::transmute::<RustOption<&mut String>, RustOption<&mut RustString>>(
+                RustOption::from(other),
+            )
         }
     }
 
     pub fn into_option_string_mut(self) -> Option<&'a mut String> {
-        let _: () = assert_option_safe::<&mut String>();
-        match self.into_value() {
-            None => None,
-            Some(r) => Some(r.as_mut_string()),
+        unsafe {
+            core::mem::transmute::<RustOption<&mut RustString>, RustOption<&mut String>>(self)
         }
+        .into_option()
     }
 
-    pub fn as_option_string_mut(&mut self) -> &mut Option<&'a mut String> {
-        let _: () = assert_option_safe::<&mut String>();
+    pub fn as_option_string_mut_mut(&mut self) -> &mut Option<&'a mut String> {
         unsafe {
-            &mut *(self as *mut RustOption<&'a mut RustString> as *mut Option<&'a mut String>)
+            core::mem::transmute::<&mut RustOption<&mut RustString>, &mut RustOption<&mut String>>(
+                self,
+            )
+            .as_mut_option()
         }
     }
 }
+
+impl<T: OptionTarget> Drop for RustOption<T> {
+    fn drop(&mut self) {
+        self.as_mut_option().take();
```

Comment:
This actually proves that assert_option_safe always works - Drop is implemented for all RustOption<T> that ever exist and I can't avoid calling it (other than MauallyDrop or equivalent I guess) so since as_mut_option calls assert_option_safe this would be a compile error if a non-safe RustOption ever was dropped anywhere (and I had to work around that at one point)